File: 0_1.baekjoon/dp/python/BOJ2624.py
from sys import stdin
import collections

readline = stdin.readline

T = int(readline())
K = int(readline())
counts = [list(map(int, readline().split())) for _ in range(K)]

def solve():
    dp =[[0 for _ in range(T+1)] for _ in range(K+1)]
    dp[0][0] = 1

    for i, (coin, limit) in enumerate(counts, 1):
        for k in range(T+1):
            dp[i][k] = dp[i-1][k]
        for count in range(1, limit+1):
            for money in range(coin*count, T+1):
                dp[i][money] += dp[i-1][money - coin * count]

    print(dp[-1][-1])

solve()

File: 0_1.baekjoon/dp/java/src/BOJ2624.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.StringTokenizer;

public class BOJ2624 {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));

        int T = Integer.parseInt(br.readLine());
        int k = Integer.parseInt(br.readLine());

        int[][] counts = new int[k][2];

        for (int i=0;i<k;i++){
            StringTokenizer st = new StringTokenizer(br.readLine(), " ");
            counts[i] = new int[] {Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())};
        }

        int[][] dp = new int[k+1][T+1];
        dp[0][0] = 1;

        for (int i=1;i<=k;i++) {
            int coin = counts[i-1][0];
            int limit = counts[i-1][1];
            for (int j=0;j<=T;j++){
                dp[i][j] = dp[i-1][j];
            }
            for (int count=1;count<=limit;count++) {
                for (int money=coin*count;money<=T;money++) {
                    dp[i][money] += dp[i-1][money - count*coin];
                }
            }
        }
        System.out.println(dp[k][T]);
    }
}
